components/data-feed-service/service/data-feed-workflow.go

```diff
@@ -0,0 +1,157 @@
+package service
+
+import (
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/chef/automate/lib/cereal"
+)
+
+var (
+	dataFeedWorkflowName = cereal.NewWorkflowName("data-feed-workflow")
+	dataFeedScheduleName = "periodic-data-feed-workflow"
+)
+
+type DataFeedWorkflowExecutor struct {
+	workflowName cereal.WorkflowName
+}
+
+// DataFeedWorkflowParams the params for the workflow and all the tasks in the workflow
+type DataFeedWorkflowParams struct {
+	UpdatedNodesOnly bool
+	NodeBatchSize    int
+	NodeIDs          map[string]NodeIDs
+	FeedStart        time.Time
+	FeedEnd          time.Time
+	PollTaskParams   DataFeedPollTaskParams
+}
+
+type DataFeedWorkflowPayload struct {
+	TasksComplete bool
+	NodeIDs       []map[string]NodeIDs
+	NodeIDsIndex  int
+}
+
+func (e *DataFeedWorkflowExecutor) OnStart(w cereal.WorkflowInstance, ev cereal.StartEvent) cereal.Decision {
+	params := DataFeedWorkflowParams{}
+	log.Debugf("OnStart params %v", params)
+	err := w.GetParameters(&params)
+	if err != nil {
+		return w.Fail(err)
+	}
+
+	/*
+	 * Start the workflow with the poll task to get any nodes that have had client runs
+	 */
+	err = w.EnqueueTask(dataFeedPollTaskName, params)
+	if err != nil {
+		return w.Fail(err)
+	}
+
+	initialPayload := DataFeedWorkflowPayload{}
+	return w.Continue(&initialPayload)
+}
+
+func (e *DataFeedWorkflowExecutor) OnTaskComplete(w cereal.WorkflowInstance, ev cereal.TaskCompleteEvent) cereal.Decision {
+	params := DataFeedWorkflowParams{}
+	err := w.GetParameters(&params)
+	if err != nil {
+		return w.Fail(err)
+	}
+
+	log.Debugf("OnTaskComplete params: %v", params)
+	payload := DataFeedWorkflowPayload{}
+	err = w.GetPayload(&payload)
+	if err != nil {
+		return w.Fail(err)
+	}
+
+	switch ev.TaskName {
+	case dataFeedPollTaskName:
+		taskResults, err := getPollTaskResults(ev)
+		if err != nil {
+			return w.Fail(err)
+		}
+		/*
+		 * Update the workflow params
+		 */
+		params.NodeIDs = taskResults.NodeIDs
```

Comment:
NodeIDs looks like it is not something the caller of the workflow passes in, and is derived while running the workflow. I found this a little confusing. It might be worth having the parameters for the task be distinct from the workflow parameters.

Reply:
I'll raise an issue to look at this on our Zen board. I am using the workflow parameters as they need to be updated with the feed times for updating the schedule. Initially I had params per task but moved away for the simplicity with the feed times. It's worth looking at again now that the tasks have been consolidated.